File: seeed_arm_console/sdk/cpp/examples/gateway_client.cpp
#include "rebot_sdk/client.hpp"

#include <cstdlib>
#include <iostream>
#include <memory>
#include <vector>

int main(int argc, char** argv) {
    const std::string gateway_address = argc > 1 ? argv[1] : "127.0.0.1:50051";
    auto channel = grpc::CreateChannel(gateway_address, grpc::InsecureChannelCredentials());
    rebot::sdk::ArmGatewayClient gateway(std::move(channel), "cpp-sdk-example");

    rebot::sdk::ConnectionInfo connection;
    auto status = gateway.handshake(&connection);
    if (!status.ok()) {
        std::cerr << "handshake failed: " << status.error_message() << '\n';
        return EXIT_FAILURE;
    }
    std::cout << "connected source=" << connection.source << " dof=" << connection.dof
              << " session=" << connection.session_id << '\n';

    rebot::sdk::CommandAck ack;
    status = gateway.enable(true, &ack);
    if (!status.ok() || !ack.accepted()) {
        std::cerr << "enable rejected: " << (status.ok() ? ack.reason : status.error_message()) << '\n';
        return EXIT_FAILURE;
    }
    status = gateway.subscribe_telemetry(
        10,
        [](const rebot::sdk::TelemetryFrame& frame) {
            std::cout << "telemetry sequence=" << frame.sequence
                      << " joints=" << frame.joint_position_rad.size() << '\n';
            return false;
        },
        5000);
    if (!status.ok() && status.error_code() != grpc::StatusCode::CANCELLED) {
        std::cerr << "telemetry failed: " << status.error_message() << '\n';
        return EXIT_FAILURE;
    }
    status = gateway.stop(false, &ack);
    if (!status.ok() || !ack.accepted()) {
        std::cerr << "stop rejected: " << (status.ok() ? ack.reason : status.error_message()) << '\n';
        return EXIT_FAILURE;
    }
    status = gateway.reset_fault(&ack);
    if (!status.ok() || !ack.accepted()) {
        std::cerr << "reset_fault rejected: " << (status.ok() ? ack.reason : status.error_message()) << '\n';
        return EXIT_FAILURE;
    }
    std::cout << "cpp_sdk=OK\n";
    return EXIT_SUCCESS;
}

File: seeed_arm_console/sdk/rust/examples/gateway_client.rs
use rebot_arm_sdk::ArmGatewayClient;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let endpoint = std::env::var("ARM_GATEWAY_GRPC_URL")
        .unwrap_or_else(|_| "http://127.0.0.1:50051".to_owned());
    let mut gateway = ArmGatewayClient::connect(endpoint, "rebot-sdk-rust-example").await?;
    let connection = gateway.handshake().await?;
    println!(
        "connected session={} source={} dof={}",
        connection.session_id, connection.source, connection.dof
    );

    let enable = gateway.enable(true, "rust-example-enable").await?;
    println!("enable={} reason={}", enable.status, enable.reason);
    let mut stream = gateway.subscribe_telemetry(20).await?;
    if let Some(frame) = stream.message().await? {
        println!(
            "telemetry sequence={} joints={} quality={}",
            frame.sequence,
            frame.joint_position_rad.len(),
            frame.quality
        );
    }
    let stop = gateway.stop(false, "rust-example-stop").await?;
    println!("stop={} reason={}", stop.status, stop.reason);
    Ok(())
}
